### PR TITLE
feat:  get config without sending file

### PR DESCRIPTION
**Pull request by eshenko**

### CODE REVIEW

**Comment by iliaonishchenko** [comment]
Coverage?

**Comment by eshenko** [comment]
Coverage is OK now.

File: repositories/src/main/scala/cromwell/pipeline/datastorage/dto/Project.scala [hunk 1]
```diff
@@ -196,11 +196,23 @@ object Visibility {
...
+  version: Option[PipelineVersion]
+)
 
 object ProjectBuildConfigurationRequest {
-  implicit val projectBuildConfigurationRequestFormat: OFormat[ProjectBuildConfigurationRequest] =
-    Json.format[ProjectBuildConfigurationRequest]
+
+  import cromwell.pipeline.datastorage.dto.ProjectFile.pathFormat
+
+  implicit val projectBuildConfigurationRequestFormat: OFormat[ProjectBuildConfigurationRequest] = {
```
Comment (myazinn) [inline repositories/src/main/scala/cromwell/pipeline/datastorage/dto/Project.scala]: Why not use macro implementation?

File: services/src/main/scala/cromwell/pipeline/service/ProjectFileService.scala [hunk 2]
```diff
@@ -32,11 +33,20 @@ class ProjectFileService(
...
+  ): Future[Either[VersioningException, ProjectConfiguration]] =
+    for {
+      project <- projectService.getUserProjectById(projectId, userId)
+      eitherFile <- projectVersioning.getFile(project, projectFilePath, version)
+    } yield eitherFile match {
+      case Right(file) =>
+        womTool.inputsToList(file.content.content) match {
+          case Right(nodes) =>
+            Right(ProjectConfiguration(projectId, active = true, List(ProjectFileConfiguration(file.path, nodes))))
+          case Left(e) => Left(VersioningException.FileException(e.head))
```
Comment (myazinn) [inline services/src/main/scala/cromwell/pipeline/service/ProjectFileService.scala]: Shouldn't we show all errors, not just the first one?

File: services/src/test/scala/cromwell/pipeline/service/ProjectFileServiceTest.scala [hunk 3]
```diff
@@ -79,22 +78,28 @@ class ProjectFileServiceTest extends AsyncWordSpec with Matchers with MockitoSug
       val projectId = project.projectId
       val projectFileContent = ProjectFileContent("File content")
       val projectFile = ProjectFile(Paths.get("test.txt"), projectFileContent)
+      val version = Some(TestProjectUtils.getDummyPipeLineVersion())
+      val userId = UserId.random
 
       "return success message for request" taggedAs Service in {
+        when(projectService.getUserProjectById(projectId, userId)).thenReturn(Future(project))
```
Comment (myazinn) [inline services/src/test/scala/cromwell/pipeline/service/ProjectFileServiceTest.scala]:
```suggestion
        when(projectService.getUserProjectById(projectId, userId)).thenReturn(Future.successful(project))
```
Same for other places below

File: repositories/src/test/scala/cromwell/pipeline/datastorage/dao/repository/utils/GeneratorUtils.scala [hunk 4]
```diff
@@ -113,8 +113,8 @@ object GeneratorUtils {
 
   lazy val projectBuildConfigurationRequestGen: Gen[ProjectBuildConfigurationRequest] = for {
     projectId <- projectIdGen
-    projectFile <- projectFileGen
-  } yield ProjectBuildConfigurationRequest(projectId, projectFile)
+    pipelineVersion <- Gen.option(pipelineVersionGen)
+  } yield ProjectBuildConfigurationRequest(projectId, Paths.get("test.wdl"), pipelineVersion)
```
Comment (myazinn) [inline repositories/src/test/scala/cromwell/pipeline/datastorage/dao/repository/utils/GeneratorUtils.scala]: use generator for `Path` as well

**Review by myazinn — CHANGES_REQUESTED** [review]

File: repositories/src/main/scala/cromwell/pipeline/datastorage/dto/Project.scala (hunk 1 above)
Comment (eshenko) [reply repositories/src/main/scala/cromwell/pipeline/datastorage/dto/Project.scala]: Fixed.

File: services/src/main/scala/cromwell/pipeline/service/ProjectFileService.scala (hunk 2 above)
Comment (eshenko) [reply services/src/main/scala/cromwell/pipeline/service/ProjectFileService.scala]: Fixed.

File: services/src/test/scala/cromwell/pipeline/service/ProjectFileServiceTest.scala (hunk 3 above)
Comment (eshenko) [reply services/src/test/scala/cromwell/pipeline/service/ProjectFileServiceTest.scala]: Fixed.

File: repositories/src/test/scala/cromwell/pipeline/datastorage/dao/repository/utils/GeneratorUtils.scala (hunk 4 above)
Comment (eshenko) [reply repositories/src/test/scala/cromwell/pipeline/datastorage/dao/repository/utils/GeneratorUtils.scala]: Fixed.

File: services/src/main/scala/cromwell/pipeline/service/ProjectFileService.scala [hunk 5]
```diff
@@ -32,11 +33,20 @@ class ProjectFileService(
...
+  ): Future[Either[Exception, ProjectConfiguration]] =
+    for {
+      project <- projectService.getUserProjectById(projectId, userId)
+      eitherFile <- projectVersioning.getFile(project, projectFilePath, version)
+    } yield eitherFile match {
+      case Right(file) =>
+        womTool.inputsToList(file.content.content) match {
+          case Right(nodes) =>
+            Right(ProjectConfiguration(projectId, active = true, List(ProjectFileConfiguration(file.path, nodes))))
+          case Left(e) => Left(ValidationError(e.toList))
```
Comment (myazinn) [inline services/src/main/scala/cromwell/pipeline/service/ProjectFileService.scala]: Is it really a validation error? Previously it was a `VersioningException.FileException`.
Why did you change error? What this error should even mean?

File: services/src/main/scala/cromwell/pipeline/service/ProjectFileService.scala [hunk 6]
```diff
@@ -32,11 +33,20 @@ class ProjectFileService(
...
-    projectFile: ProjectFile
-  ): Future[Either[ValidationError, ProjectConfiguration]] =
-    Future(womTool.inputsToList(projectFile.content.content)).map {
-      case Left(value) => Left(ValidationError(value.toList))
-      case Right(nodes) =>
-        Right(ProjectConfiguration(projectId, active = true, List(ProjectFileConfiguration(projectFile.path, nodes))))
+    projectFilePath: Path,
+    version: Option[PipelineVersion],
+    userId: UserId
+  ): Future[Either[Exception, ProjectConfiguration]] =
```
Comment (myazinn) [inline services/src/main/scala/cromwell/pipeline/service/ProjectFileService.scala]: I just noticed that return type is `Future[Either[Exception, _]]`

I think you should use `Future[_]` instead. `Future[Either[_, _]]` makes sense when you want to explicitly show that you expect some specific errors. But `Future[Either[Exception, _]]` is almost as vague as `Future[Either[Throwable, _]]` which can be simplified to `Future[_]`.

Comment (eshenko) [reply services/src/main/scala/cromwell/pipeline/service/ProjectFileService.scala]: Fixed.

File: services/src/main/scala/cromwell/pipeline/service/ProjectFileService.scala (hunk 5 above)
Comment (eshenko) [reply services/src/main/scala/cromwell/pipeline/service/ProjectFileService.scala]: womTool returns NonEmptyList of validation errors, now I concatenate all errors to one string and throw them as ValidationExceprion.

VersioningException was my mistake.

File: controllers/src/main/scala/cromwell/pipeline/controller/ProjectFileController.scala [hunk 7]
```diff
@@ -65,10 +66,16 @@ class ProjectFileController(wdlService: ProjectFileService)(implicit val executi
...
-              case Failure(e)                    => complete(StatusCodes.InternalServerError, e.getMessage)
+            onComplete(
+              wdlService
+                .buildConfiguration(request.projectId, request.projectFilePath, request.version, accessToken.userId)
+            ) {
+              case Success(configuration) => complete(configuration)
+              case Failure(e) =>
+                e match {
+                  case ValidationException(msg) => complete(StatusCodes.UnprocessableEntity, msg)
+                  case _                        => complete(StatusCodes.InternalServerError, e.getMessage)
```
Comment (myazinn) [inline controllers/src/main/scala/cromwell/pipeline/controller/ProjectFileController.scala]:
```suggestion
              case Failure(ValidationException(msg)) => complete(StatusCodes.UnprocessableEntity, msg)
              case Failure(e) => complete(StatusCodes.InternalServerError, e.getMessage)
```

In this case it's better to have a few duplicated symbols rather than nested nested PM.

File: repositories/src/main/scala/cromwell/pipeline/datastorage/dto/ValidationError.scala [hunk 8]
```diff
@@ -1,3 +1,4 @@
 package cromwell.pipeline.datastorage.dto
 
 final case class ValidationError(errors: List[String]) extends Exception
+final case class ValidationException(message: String) extends Exception(message)
```
Comment (myazinn) [inline repositories/src/main/scala/cromwell/pipeline/datastorage/dto/ValidationError.scala]: What is the difference between them?

File: services/src/main/scala/cromwell/pipeline/service/ProjectFileService.scala [hunk 9]
```diff
@@ -32,11 +33,20 @@ class ProjectFileService(
...
+      project <- projectService.getUserProjectById(projectId, userId)
+      eitherFile <- projectVersioning.getFile(project, projectFilePath, version)
+    } yield eitherFile match {
+      case Right(file) =>
+        womTool.inputsToList(file.content.content) match {
+          case Right(nodes) =>
+            ProjectConfiguration(projectId, active = true, List(ProjectFileConfiguration(file.path, nodes)))
+          case Left(e) => throw ValidationException(e.toList.mkString(", "))
+        }
+      case Left(versioningException) => throw versioningException
```
Comment (myazinn) [inline services/src/main/scala/cromwell/pipeline/service/ProjectFileService.scala]: In Scala we don't throw exceptions. Especially when method's return type allows us to fail gracefully.

**Review by myazinn — CHANGES_REQUESTED** [review]

File: repositories/src/main/scala/cromwell/pipeline/datastorage/dto/ValidationError.scala (hunk 8 above)
Comment (eshenko) [reply repositories/src/main/scala/cromwell/pipeline/datastorage/dto/ValidationError.scala]: The same. Removed:)

File: services/src/main/scala/cromwell/pipeline/service/ProjectFileService.scala (hunk 9 above)
Comment (eshenko) [reply services/src/main/scala/cromwell/pipeline/service/ProjectFileService.scala]: Fixed.

File: controllers/src/main/scala/cromwell/pipeline/controller/ProjectFileController.scala (hunk 7 above)
Comment (eshenko) [reply controllers/src/main/scala/cromwell/pipeline/controller/ProjectFileController.scala]: Fixed.

File: controllers/src/main/scala/cromwell/pipeline/controller/ProjectFileController.scala [hunk 10]
```diff
@@ -65,9 +66,12 @@ class ProjectFileController(wdlService: ProjectFileService)(implicit val executi
       path("files" / "configurations") {
         post {
```
Comment (myazinn) [inline controllers/src/main/scala/cromwell/pipeline/controller/ProjectFileController.scala]: Shouldn't we use `get` instead? We are not actually posting anything, we just want to get the configuration of the existing file.

File: services/src/main/scala/cromwell/pipeline/service/ProjectFileService.scala [hunk 11]
```diff
@@ -32,11 +33,26 @@ class ProjectFileService(
...
-    Future(womTool.inputsToList(projectFile.content.content)).map {
-      case Left(value) => Left(ValidationError(value.toList))
-      case Right(nodes) =>
-        Right(ProjectConfiguration(projectId, active = true, List(ProjectFileConfiguration(projectFile.path, nodes))))
+    projectFilePath: Path,
+    version: Option[PipelineVersion],
+    userId: UserId
+  ): Future[ProjectConfiguration] = {
+
+    val either = for {
```
Comment (myazinn) [inline services/src/main/scala/cromwell/pipeline/service/ProjectFileService.scala]:
```suggestion
    val eitherFile = for {
```

**Review by myazinn — CHANGES_REQUESTED** [review]
Sorry that I didn't notice major problem with REST method earlier

File: services/src/main/scala/cromwell/pipeline/service/ProjectFileService.scala (hunk 11 above)
Comment (eshenko) [reply services/src/main/scala/cromwell/pipeline/service/ProjectFileService.scala]: Fixed.

File: controllers/src/main/scala/cromwell/pipeline/controller/ProjectFileController.scala (hunk 10 above)
Comment (eshenko) [reply controllers/src/main/scala/cromwell/pipeline/controller/ProjectFileController.scala]: It's not my changing, so maybe it's out of scope this PR?

Comment (myazinn) [reply controllers/src/main/scala/cromwell/pipeline/controller/ProjectFileController.scala]: It **is** your change. You've significantly changed this endpoint and now it works differently and accept other parameters. It's reasonable to change HTTP method to a more suitable one.

Comment (eshenko) [reply controllers/src/main/scala/cromwell/pipeline/controller/ProjectFileController.scala]: Got it thanks.
Fixed.

File: controllers/src/main/scala/cromwell/pipeline/controller/ProjectFileController.scala [hunk 12]
```diff
@@ -63,13 +66,21 @@ class ProjectFileController(wdlService: ProjectFileService)(implicit val executi
...
-              case Failure(e)                    => complete(StatusCodes.InternalServerError, e.getMessage)
-            }
+        get {
+          parameters('project_id.as[String], 'project_file_path.as[String], 'version.as[String].optional) {
+            (projectId, projectFilePath, version) =>
+              onComplete(
+                wdlService.buildConfiguration(
+                  ProjectId(projectId),
+                  Paths.get(projectFilePath),
+                  version.map(v => PipelineVersion(v)),
```
Comment (myazinn) [inline controllers/src/main/scala/cromwell/pipeline/controller/ProjectFileController.scala]: Let's make this deserialization reusable. We can do something like this
```
parameters('project_id.as[ProjectId], 'project_file_path.as[Path], 'version.as[PipelineVersion].optional)
```
We just need to tell Akka how to deserialize our classes. Take a look [here](https://github.com/EpamLifeSciencesTeam/cmwl_pipeline/blob/4fd247669f7118b967306ca2776340d7b0ed22ab/controllers/src/main/scala/cromwell/pipeline/controller/RunController.scala#L24) and [here](https://github.com/EpamLifeSciencesTeam/cmwl_pipeline/blob/4fd247669f7118b967306ca2776340d7b0ed22ab/controllers/src/main/scala/cromwell/pipeline/controller/utils/FromStringUnmarshallers.scala#L9). In this case everyone will be able to reuse it without boilerplate.

**Review by myazinn — CHANGES_REQUESTED** [review]

File: controllers/src/main/scala/cromwell/pipeline/controller/ProjectFileController.scala (hunk 12 above)
Comment (eshenko) [reply controllers/src/main/scala/cromwell/pipeline/controller/ProjectFileController.scala]: Thanks. Fixed.

**Review by GrigoriiBerezin — APPROVED** [review]
Nice work! Also thanks for changing postman files

**Review by myazinn — APPROVED** [review]
LGTM, but now you have a merge conflict. Please resolve it once you have free time.

**Comment by eshenko** [comment]
> LGTM

Conflicts resolved.
